### PR TITLE
PBTree Flush Control Monitor

### PR DESCRIPTION
**Pull request by Cpaulyz**
### Content3 ...

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/metric/SchemaEngineCachedMetric.java [hunk 1]
```diff
@@ -109,15 +100,15 @@ public void bindTo(AbstractMetricService metricService) {
...
+        ReleaseFlushMonitor::getActiveWorkerNum,
         Tag.NAME.toString(),
         RELEASE_THREAD_NUM);
     metricService.createAutoGauge(
         Metric.SCHEMA_ENGINE.toString(),
         MetricLevel.IMPORTANT,
-        CacheMemoryManager.getInstance(),
-        CacheMemoryManager::getFlushThreadNum,
+        ReleaseFlushMonitor.getInstance(),
+        ReleaseFlushMonitor::getActiveWorkerNum,
```
Comment (Cpaulyz) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/metric/SchemaEngineCachedMetric.java:111]: Merge these two metrics

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    return scheduler.getActiveWorkerNum();
+  }
+
+  private ReleaseFlushMonitor() {}
+
+  private static class GlobalCacheManagerHolder {
+    private static final ReleaseFlushMonitor INSTANCE = new ReleaseFlushMonitor();
+
+    private GlobalCacheManagerHolder() {}
+  }
```
Comment (MarcosZyk) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java:272]: Rename this.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            throw new NoSuchElementException();
+          }
+          cur = cur.next;
+          return cur;
+        }
+
+        @Override
+        public void remove() {
+          RecordList.this.remove(cur);
+        }
```
Comment (MarcosZyk) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java:330]: Might cause NPE if the cur is removed during iterating.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/flush/Scheduler.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                                      lockManager);
+                              try {
+                                flushExecutor.flushVolatileNodes();
+                              } catch (MetadataException e) {
+                                LOGGER.warn(
+                                    "Error occurred during MTree flush, current SchemaRegionId is {} because {}",
+                                    regionId,
+                                    e.getMessage(),
+                                    e);
+                              } finally {
```
Comment (MarcosZyk) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/flush/Scheduler.java:138]: What about encapsulate such logic into PBTreeFlushExecutor?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/flush/Scheduler.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+            try {
+              IDatabaseMNode<ICachedMNode> dbNode = cacheManager.collectUpdatedStorageGroupMNodes();
+              if (dbNode != null) {
+                flushExecutor = new PBTreeFlushExecutor(dbNode, cacheManager, file, lockManager);
+                flushExecutor.flushDatabase();
+                remainToFlush.decrementAndGet();
+              }
+              Iterator<ICachedMNode> volatileSubtrees = cacheManager.collectVolatileSubtrees();
+              while (volatileSubtrees.hasNext()) {
+                nodesToFlush.add(volatileSubtrees.next());
```
Comment (MarcosZyk) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/flush/Scheduler.java:219]: Need more discussion here. 

Such behavior may lock all the collected node, and none R/W operation will be allowed to be executed under the related subtree until the subtree has been flushed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java (hunk 2 above)
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java:272]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java (hunk 3 above)
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java:330]: Actually there are only one thread using this iterator... NPE will not be occurred without concurrency. So is it ok to add @NotThreadSafe?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/flush/Scheduler.java (hunk 5 above)
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/flush/Scheduler.java:219]: It seems that encapsulating the logic of taking the node out of the iterator in the PBTreeFlushExecutor would solve the problem of holding locks. 

But there will be a new problem: Schedule will have a counter for restoreToFlush to record the subtreeRoots of each timed task flush. Ff the logic is encapsulated, there is no way for the Scheduler to sense it in advance. Can the CacheManager give an interface to estimate the number of subtreeRoots in an iterator?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java (hunk 3 above)
Comment (MarcosZyk) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java:330]: I will correct my description. After removal, the iterator is bound to return null. Is this expected?

Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/ReleaseFlushMonitor.java:330]: Fixed

**Pull request merged**